tests/forms/test_address_upload.py

```diff
@@ -0,0 +1,170 @@
+    assert response.status_code == 200
+    assert b'Syntax error on line 1:' in response.content
+
+
+@pytest.mark.django_db
+def test_file_upload_length_long(client, organizer, event, env):
+    client.login(email='admin@pretix', password='admin')
+
+    request_url = f'/control/event/{organizer.slug}/{event.slug}/wallet-address-upload/'
+    invalid_file = SimpleUploadedFile('upload.txt', """
+        0x38A670EB58F9D63Dfds32D279D92dC651155595e66009
+    """.encode("utf-8"), content_type="text/plain")
+
+    response = client.post(request_url, {'wallet_addresses': invalid_file}, follow=True)
+    assert response.status_code == 200
+    assert b'Syntax error on line 1:' in response.content
+
+
+@pytest.mark.django_db
+def test_file_upload_one_address_per_line(client, organizer, event, env):
+    client.login(email='admin@pretix', password='admin')
+
+    request_url = f'/control/event/{organizer.slug}/{event.slug}/wallet-address-upload/'
+    invalid_file = SimpleUploadedFile('upload.txt', """
+        # Two addresses in the same line, expected one
+        0xe61407b0708CC10006aAc0ceA62F553Ed84D2aD8 0xcD9989Dc8F0f02866eC945cBCCDd6CFA32D46026
+    """.encode("utf-8"), content_type="text/plain")
+
+    response = client.post(request_url, {'wallet_addresses': invalid_file}, follow=True)
+    assert response.status_code == 200
+    assert b'Syntax error on line 1:' in response.content
+
+
+@pytest.mark.django_db
+def test_file_upload_invalid_middle(client, organizer, event, env):
+    client.login(email='admin@pretix', password='admin')
+
+    request_url = f'/control/event/{organizer.slug}/{event.slug}/wallet-address-upload/'
+    invalid_file = SimpleUploadedFile('upload.txt', """
+        # Invalid address in the middle of the file
+        0xe61407b0708CC10006aAc0ceA62F553Ed84D2aD8
+        0x378c3Ba28099187788500cb4a833FbA161F9879C
+        0xcD9989Dc8F0f02866eC9BCCDd6CFA32D46026
+        0x1F59154Bcf62A3B63Db28F306996540c0a6aD8a5
+    """.encode("utf-8"), content_type="text/plain")
+
+    response = client.post(request_url, {'wallet_addresses': invalid_file}, follow=True)
+    assert response.status_code == 200
+    assert b'Syntax error on line 3:' in response.content
+
+
+@pytest.mark.django_db
+def test_file_upload_invalid_prefix(client, organizer, event, env):
+    client.login(email='admin@pretix', password='admin')
+
+    request_url = f'/control/event/{organizer.slug}/{event.slug}/wallet-address-upload/'
+    invalid_file = SimpleUploadedFile('upload.txt', """
+        # Prefix is not 0x
+        e61407b0708CC10006aAc0ceA62F553Ed84D2aD8
+    """.encode("utf-8"), content_type="text/plain")
+
+    response = client.post(request_url, {'wallet_addresses': invalid_file}, follow=True)
+    assert response.status_code == 200
+    assert b'Syntax error on line 1:' in response.content
+
+
+@pytest.mark.django_db
+def test_file_upload_empty_file(client, organizer, event, env):
+    client.login(email='admin@pretix', password='admin')
```

Comment:
*nitpick*

Not necessary, but since this is repeated in each of the tests, you *could* have a fixture like this:

```python
@pytest.fixture
def admin_client(client):
    client.login(email='admin@pretix', password='admin')
    return client
```

But as I'm writing this I realize that is sort of a [foot gun](https://en.wiktionary.org/wiki/footgun) because someone *might* write a test like this:

```python
def test_something(client, admin_client):
    # ...
```

In which case they wouldn't realize that both `client` and `admin_client` are the same object and are both logged in as the admin user...  So maybe go a step further and create a new client instance if you are planning to do this?  I'm not actually sure it's a lot better, but it does remove a  little bit of duplicated code at the added cost of extra fixtures and abstraction layers